### DISCUSSION TITLE
Derivative Order

### DISCUSSION THREAD

**Discussion by ejmeitz** [brianguenter/FastDifferentiation.jl#64] [Q&A] [answered]
For the `derivative` function which order are the derivatives take in?

If I have derivative([f], x, y, z) is the derivative w.r.t x or z take first? Based on the doc string it kind of looks like z would be first. In my mind x would get taken first. I know this doesn't actually affect the result for a single derivative but I think its messing up a higher dimensional array of derivatives I was building. (Not 100% sure this is the culprit I'm still checking)

**Comment by brianguenter** [comment]
I looked at the code for `derivative`. It should compute the derivative for x first, then y, then z.

If you have a MWE for your problem I might be able to help you figure out what the problem is.

In any case a doc PR would be good because this isn't made explicit.

**Reply by ejmeitz** [reply]
I think I figured out my problem it was just the ordering. I'll make a PR to specify this though as the order is the opposite of what I expected based on the docstring.

**Comment by ejmeitz** [comment]
Actually something weird is still going on.

I am calculating a 3x3x3 block of derivatives of the form for a given (i,j,k):
![image](https://github.com/brianguenter/FastDifferentiation.jl/assets/54505069/46fa80dc-10c3-4212-9f16-8fb96dbba398)
Here the $\alpha$, $\beta$, $\gamma$ represent cartesian directions (x,y,z). So $r_i^{x}$ is the x component of this vector. I calculate a block of these derivatives for fixed (i,j,k) with this code:

```julia
r_i_vars = make_variables(:ri, 3)
r_j_vars = make_variables(:rj, 3)
r_k_vars = make_variables(:rk, 3)

r_ij_threebody = r_i_vars .- r_j_vars
r_ik_threebody = r_i_vars .- r_k_vars
r_ij_norm_threebody = sqrt(sum(x -> x^2, r_ij_threebody))
r_ik_norm_threebody = sqrt(sum(x -> x^2, r_ik_threebody))

H3_symbolic_iij = Array{FastDifferentiation.Node}(undef, 3, 3, 3)

pot3_symbolic = three_body_potential_nounits(pot, r_ij_threebody, r_ik_threebody, r_ij_norm_threebody, r_ik_norm_threebody)

for alpha in range(1,3)
        for beta in range(1,3)
            for gamma in range(1,3)
                  H3_symbolic_iij[alpha,beta,gamma] = derivative([pot3_symbolic], r_j_vars[gamma], r_i_vars[beta], r_i_vars[alpha])[1]
             end
        end
end

H3_exec_iij = make_function(H3_symbolic_iij, [r_i_vars; r_j_vars; r_k_vars])
```

The issue I'm seeing is that when I evaluate this derivative with (i,j,k) = (1,1,2) I am trying to use `permutedims` to get the (1,2,1) and (2,1,1) versions of this tensor.

If I want the 121 block the permutation order should not matter as long as the derivative w.r.t 2 (axis 3) ends up in the middle. But that is not true here. I checked that this should hold with some code that just prints strings to represent the derivatives (also below).

```julia
block = H3_exec_iij([r1;r1;r2]) #for some arbitrary r1, r2 which are 3x1 vectors
permutedims(block , (2,3,1)) == permutedims(block , (1,3,2)) #FALSE


function make_pattern(a,b,c)
    out = Array{Vector{String}}(undef, (3,3,3))
    for (i,x) in enumerate([:x,:y,:z])
        for (j,y) in enumerate([:x,:y,:z])
            for (k,z) in enumerate([:x,:y,:z])
                out[i,j,k] = sort(["r_$(a)$(x)", "r_$(b)$(y)", "r_$(c)$(z)"])
            end
        end
    end
    return out
end

 block_112 = make_pattern(1,1,2)
permutedims(block_112, (2,3,1)) == permutedims(block_112, (1,3,2)) #TRUE

```

Don't really think this is an issue with the library, I'm probably being dumb

**Comment by brianguenter** [comment]
Your code is not executable as is (undefined variables and functions) so I can't tell if there is something wrong with FastDifferentiation or if it's something elsewhere in your code. If you come to the conclusion that your code is correct and the error lies in FastDifferentiation then give me a full MWE and I will track the problem down.

**Reply by ejmeitz** [reply]
Im making a MWE right now. I'm not sure its a bug or just me interpreting the FastDiff interface wrong.

Will post the MWE in a bit.

**Comment by ejmeitz** [comment]
Ok when I use a simpler function this property holds, but for some reason in my code this test fails. I've attached that file with the MWE that works and a MWE with my code that does not. The only difference is in the function I differentiate which should not matter for this property to hold.

You'll need to `import FastDifferentiation, LinearAlgebra, Unitful` for these. You can use 
`r_test = [4.0725, 1.3575, 9.5025, 2.715, 0.0, 8.145, 5.43, 2.715, 8.145]` as the test point.

[MWE_working.zip](https://github.com/brianguenter/FastDifferentiation.jl/files/14711475/MWE_working.zip)
[MWE_failing.zip](https://github.com/brianguenter/FastDifferentiation.jl/files/14711476/MWE_failing.zip)

**Comment by brianguenter** [comment]
For your `MWE_working` example it looks like both the iij and iji derivatives are identically zero, so of course they will be equal. Do you get the same result on your end?

I modified your example slightly to print out the symbolic derivatives:

```julia
function MWE_working(r_test)

    D = 3
    @assert length(r_test) == 9

    r_i_vars = make_variables(:ri, D)
    r_j_vars = make_variables(:rj, D)
    r_k_vars = make_variables(:rk, D)

    H3_symbolic_iij = Array{FastDifferentiation.Node}(undef, D, D, D)
    H3_symbolic_iji = Array{FastDifferentiation.Node}(undef, D, D, D)

    function my_norm(r_arr)
        return sqrt(sum(x -> x^2, r_arr))
    end

    function f(r_arr)
        return my_norm(r_arr[1:3]) + my_norm(r_arr[4:6]) + my_norm(r_arr[7:9])
    end

    f_symbolic = f([r_i_vars; r_j_vars; r_k_vars])

    for a in range(1, D)
        for b in range(1, D)
            for c in range(1, D)
                H3_symbolic_iij[a, b, c] = derivative([f_symbolic], r_j_vars[c], r_i_vars[b], r_i_vars[a])[1]
                H3_symbolic_iji[a, b, c] = derivative([f_symbolic], r_i_vars[c], r_j_vars[b], r_i_vars[a])[1]
            end
        end
    end

    println("iij $H3_symbolic_iij")
    println("iji $H3_symbolic_iji")

    r_vars = [r_i_vars; r_j_vars; r_k_vars]
    H3_exec_iij = make_function(H3_symbolic_iij, r_vars)
    H3_exec_iji = make_function(H3_symbolic_iji, r_vars)

    block_iij = H3_exec_iij(r_test)
    block_iji = H3_exec_iji(r_test)

    test1 = block_iij == permutedims(block_iji, (1, 3, 2))
    test2 = block_iij == permutedims(block_iji, (2, 3, 1))

    return test1, test2

end
export MWE_working
```
result of running `MWE_working`:

```julia
julia> MWE_working(r_test)
iij FastDifferentiation.Node[0.0 0.0 0.0; 0.0 0.0 0.0; 0.0 0.0 0.0;;; 0.0 0.0 0.0; 0.0 0.0 0.0; 0.0 0.0 0.0;;; 0.0 0.0 0.0; 0.0 0.0 0.0; 0.0 0.0 0.0]
iji FastDifferentiation.Node[0.0 0.0 0.0; 0.0 0.0 0.0; 0.0 0.0 0.0;;; 0.0 0.0 0.0; 0.0 0.0 0.0; 0.0 0.0 0.0;;; 0.0 0.0 0.0; 0.0 0.0 0.0; 0.0 0.0 0.0]
(true, true)
```

**Reply by ejmeitz** [reply]
I apologize I was rushing at the end of the day, yes the working MWE I sent you was all zeros. I've modified the code below to not give zeros, with the tests still passing. I also changed some stuff in the failing case.

[MWEs_fixed.zip](https://github.com/brianguenter/FastDifferentiation.jl/files/14723937/MWEs_fixed.zip)

I'm not sure what's different about my function, it is basically just a complicated combination of `norm`, `exp`, and `dot`. Derivative order should not matter, but if you look at the numbers produced by the MWE_failing they are completely different. Its not just the order in the array that's being changed with my code.

**Comment by brianguenter** [comment]
It could be a bug in FastDifferentiation, unrelated to derivative order. There are several of these still unresolved - one of them might be affecting your results. These bugs typically only manifest in large expressions with complex graphs (which makes them devilishly hard to debug).

Your function doesn't seem that big so it would be a little surprising if this was the case. 

A simple way to check FastDifferentiation results is to use FiniteDifferences.jl to compute derivatives. Use BigFloats as inputs so roundoff error will not be significant and use a high order finite difference approximation, say 20th or 30th order.

If FiniteDifferences gives you the same numerical results as FastDifferentiation then the error is probably something to do with a misinterpretation of the order in which `derivative` computes derivatives. 

If FiniteDifferences doesn't match FastDifferentiation then it is probably one of the unresolved bugs. Or maybe a brand new one.

Alternatively, you could try using Symbolics.jl to compute the symbolic derivative and generate an executable, then compare the results to FastDifferentiation.

**Reply by ejmeitz** [reply]
I'll try those and see what happens. I checked the second order version of this with FiniteDifferences.jl awhile back and those were alright but going up one more derivative could be an issue. Will let you know what I find.

Thanks!

**Reply by ejmeitz** [reply]
@brianguenter not sure if this helps with your debugging, but removing the 
```julia
cosθᵢⱼₖ = dot(rᵢⱼ, rᵢₖ) / (dist_ij * dist_ik)
# cosTheta = ((rᵢⱼ[1]*rᵢₖ[1]) + (rᵢⱼ[2]*rᵢₖ[2]) + (rᵢⱼ[3]*rᵢₖ[3]))/(dist_ij*dist_ik)
```
and just setting it equal to one removes the issues I was seeing. I also tried writing out the dot product explicitly since I had to do that for `norm` but that did not seem to matter.

The issue seems related to the dot product. If I use just ` 1/(dist_ij*dist_ik)` things work as I expect. But if I use just the dot product or the original things stop working.

Edit: I just tested Symbolics.jl and their derivatives act as I expect (they are extremely slow to compile though). Interestingly the results for `derivative([f_symbolic], r_j_vars[c], r_i_vars[b], r_i_vars[a])` from FastDiff match Symbolics, but the results for `derivative([f_symbolic], r_i_vars[c], r_j_vars[b], r_i_vars[a])` do not.

Edit 2: Some of the derivatives also just do not match Symbolics at all in any order. In this case
`H3_symbolic_iik[a,b,c] = derivative([pot3_symbolic], r_k_vars[c], r_i_vars[b], r_k_vars[a])[1]`
seems to be the only one that is always wrong.

**Comment by brianguenter** [comment]
I haven't forgotten about this problem. Can you make an even more minimal working example? The derivative graphs of these functions are gigantic, which makes it extremely hard to pinpoint the problem. Here's one of the 27 graphs for `H3_symbolic_iij`:
<img width="2003" alt="image" src="https://github.com/brianguenter/FastDifferentiation.jl/assets/116222085/58f23240-5716-4dc6-a722-e2775916f523">

**Reply by ejmeitz** [reply]
That is quite the graph lol.

I managed to boil down the expression quite a bit and also reduce the derivative to just a second derivative. For the code below I get all zeros for one order and -1 on the diagonal for the other order.  Unless I suck at algebra the case with -1 is correct.

You can also try `f = dot(r_ij, r_ik)^2` if you need something that shouldn't just be zeros. If you make `f = (r_ij_norm * r_ik_norm)^2` the function it works fine, the issue is always in the dot product from my experience. Hopefully this has small enough of a graph for you to figure things out! 

```julia
using FastDifferentiation
using LinearAlgebra

D = 3

r_i_vars = make_variables(:ri, D)
r_j_vars = make_variables(:rj, D)
r_k_vars = make_variables(:rk, D)

r_ij = r_i_vars .- r_j_vars
r_ik = r_i_vars .- r_k_vars
r_ij_norm = sqrt(sum(x -> x^2, r_ij))
r_ik_norm = sqrt(sum(x -> x^2, r_ik))


H2_symbolic_ij = Array{FastDifferentiation.Node}(undef, D, D)
H2_symbolic_ji = Array{FastDifferentiation.Node}(undef, D, D)

#f = dot(r_ij, r_ik)^2
f = dot(r_ij, r_ik)

for a in range(1,D)
    for b in range(1,D)
        H2_symbolic_ij[a,b] = derivative([f], r_i_vars[a], r_j_vars[b])[1]
        H2_symbolic_ji[a,b] = derivative([f], r_j_vars[b], r_i_vars[a])[1]
    end
end

r_vars = [r_i_vars; r_j_vars; r_k_vars]
r_test = [4.0725, 1.3575, 9.5025, 2.715, 0.0, 8.145, 5.43, 2.715, 8.145]

H2_exec_ij = make_function(H2_symbolic_ij, r_vars);
H2_exec_ji = make_function(H2_symbolic_ji, r_vars);

block_ij = H2_exec_ij(r_test)
block_ji = H2_exec_ji(r_test)

test = all(block_ij .≈ block_ji)
```

**Reply by brianguenter** [reply]
found the bug. It was an algebraic simplification rule that was incorrect. I'll fix it tonight or tomorrow and make a patch release. Check it out and see if it works for you.

**Reply by ejmeitz** [reply]
Amazing, thanks!

**Reply by brianguenter** [reply, marked as answer]
The fix is done and the patch has been submitted to the julia registry. If you download the latest version 0.3.10 of FastDifferentiation it should work for your function. If something still doesn't work let me know.

**Reply by ejmeitz** [reply]
Everything looks good! Thanks for fixing this so fast. My current fix with Symbolics was painfully slow.

**Reply by ejmeitz** [reply]
Hm I just checked the MWE's I sent, but my full code has a new error building the other derivatives:

 `AssertionError: Should only be one path from root 2 to variable 1. Instead have 2 children from node 93 on the path`

Is this that persistent issue where FastDiff cannot factor large graphs? Before this patch it was able to calculate these other derivatives fine.

**Comment by brianguenter** [comment]
Yes, unfortunately that is probably the persistent issue with larger graphs. It's the next bug on my list to fix.

Do all of your MWE's work now but there is another function that is failing? Is the one that is failing small enough that I could analyze it?
